netsim/extra/bonding/plugin.py

```
@@ -0,0 +1,90 @@
+'''
+post_transform hook
+
+Apply plugin config to nodes with interfaces marked with 'bonding.ifindex', for devices that support this plugin
+'''
+def post_transform(topology: Box) -> None:
+# def post_link_transform(topology: Box) -> None: # Use 'pre_node_transform' such that only 1 IP address gets assigned, nothing to move?
+  global _config_name
+  bond_mode = topology.get('bonding.mode','active-backup')
+  bonds : Box = data.get_empty_box()                   # Map of bonds per node, indexed by bonding.ifindex
+  for node in topology.nodes.values():
+    features = devices.get_device_features(node,topology.defaults)
+    if 'bonding' in features:
+      for intf in node.get('interfaces',[]):
+        bond_ifindex = intf.get('bonding.ifindex',None)
+        if not bond_ifindex:
+          continue
+
+        link = topology.links[ intf.linkindex-1 ]
+        if 'virtual_interface' in intf or link.node_count!=2:
+          log.error( f"{intf.name}: 'bonding.ifindex' can only be applied to interfaces on direct p2p links",
+                     category=log.IncorrectAttr,module=_config_name)
+          continue
+
+        clone = data.get_box(intf)
+        if node.name in bonds and bond_ifindex in bonds[node.name]:
+          bonds[node.name][bond_ifindex]['members'].append( clone.ifname )
+          for att in ['ipv4','ipv6']:
+            intf.pop(att,None)
+        else:
+          mode = intf.get('bonding.mode',bond_mode)
+          bonds[node.name][bond_ifindex] = { 'interfaces': intf.neighbors, 'members': [ clone.ifname ], 'mode': mode }
+          for att in ['ipv4','ipv6']:                  # Move any ips (from first member link)
+            if att in intf:
+              bonds[node.name][bond_ifindex][att] = intf.pop(att,None)
+
+        if intf.get('bonding.primary',False):
+          bonds[node.name][bond_ifindex]['primary'] = intf.ifname
+
+        intf.neighbors = [ { 'ifname': i.ifname, 'node': i.node } for i in link.interfaces if i.node!=node.name ]
```

Is this "clean up the neighbor attributes" (if so, why?) or "avoid links to myself" (in which case you should report an error)

---

Yes, the cleanup is needed because the bond now gets all the indirect neighbors - like vlan peers at other hosts - while the bond member links only get directly attached peers on the link